man/man3/mfb.3
.\	/*
.\	/*	mfb.man: manual for mfb package
.\	/*	Copyright -c- 1982 Giles C. Billingsley
.\	/*
.\	/*	sccsid "%W%  %G%"
.\	/*
.tr ||
.TH MFB 3 6/21/83
.UC 4
.SH NAME
mfb \- model frame buffer interface
.SH SYNOPSIS
.B "#include <~cad/include/mfb.h>"
in the program source.
.br
.B "cc "
[ flags ] files 
.B " ~cad/lib/mfb.a -lm  "
[ libraries ]
.sp
.SH DESCRIPTION
.PP
These routines provide the user with a virtual graphics interface.
They perform the terminal dependent task of encoding\/decoding graphics
code, thereby allowing the user to write graphics programs to run on
almost any graphics device.
.PP
The user opens and initializes a graphics device by calling the
.I MFBOpen
routine that returns a pointer to that device's
.I MFB
data structure defined at the end of this manual.  By maintaining several
.I MFB
data structures, an application program can drive several graphics
devices simultaneously.  Once opened, 
.I MFB
graphics routines can be called to draw geometries, draw graphics text, 
set device parameters, or receive keyboard input.
An application program can also use any of the several utility routines that
perform line clipping, polygon clipping, or window\/viewport transformations.
Control of the graphics device is released by calling the
.I MFBClose
routine.
.PP
All programs that use
.I MFB
routines must include the file
.I mfb.h
that defines the
.I MFB
data structure to contain the information provided by
.I MFBCAP(5).
.sp
.ne 8
.SH INITIALIZATION ROUTINES
.nf
.PP
.B "MFB *MFBOpen(TerminalName, DeviceName, errorcode)"
.B "char *TerminalName, *DeviceName;"
.B "int *errorcode;"
.fi
.DT
.RS
.I MFBOpen
initializes the graphics device and fills the
.I MFB
data structure with information found in
.I MFBCAP(5).
.I TerminalName
is a pointer to a null terminated string containing the name of
the graphics device as defined in the
.I mfbcap
database file.  This argument has no default and can never be null.
.I DeviceName
is a pointer to a null terminated string containing the full path name
to the respective graphics device.  
If null,
.I stdin
and
.I stdout
are used by default.
.I errorcode
is a diagnostic integer returned by
.I MFBOpen.
The possible returned values for
.I errorcode
are defined in the
.I mfb.h
file as follows:
.PP
.ne 5
.nf
.ta \w'XXXXXXXX'u +\w'XXXXXXXXXXXXXXX'u +\w'XXXXXX'u
#define	MFBOK	 1	/* successful return */
#define	MFBBADENT	-10	/* Unknown terminal type */
#define	MFBBADMCF 	-20	/* Can't open MFBCAP file */
#define	MFBMCELNG 	-30	/* MFBCAP entry too long */
#define	MFBBADMCE 	-40	/* Bad MFBCAP entry */
#define	MFBINFMCE 	-50	/* infinite loop in MFBCAP entry */
#define	MFBBADTTY 	-60	/* stdout not in /dev */
#define	MFBBADDEV	-180	/* Can't open or close device */
#define	MFBBADOPT	-190	/* Can't access or set device stat */
#define	MFBBADWRT 	-220	/* Error during write */
.fi
.DT
.PP
Only MFBOK is not a fatal error.
.RE
.sp
.ne 8
.nf
.PP
.B "void SetCurrentMFB(mfb)"
.B "MFB *mfb;"
.fi
.DT
.RS
.I SetCurrentMFB
allows the application program to define the current graphics device.
All subsequent calls to
.I MFB
routines will affect the specified device.  Because each
.I MFB
data structure contains a separate output buffer, it is not necessary
to flush the output before resetting the current output device.
.I MFBOpen
returns with the opened graphics device defined as the current output device.
.RE
.sp
.ne 8
.nf
.PP
.B "int MFBInitialize()"
.fi
.DT
.RS
.I MFBInitialize
will flush the output buffer and (re)initialize the device for graphics input.
The graphics device or standard input will 
be placed in CBREAK mode.  See the manual
.I tty(4).
MFBOK is returned if the device was successfully initialized;
MFBBADOPT is returned if an error was encountered while attempting
to access or set the device status, and
MFBBADTTY is returned if standard output can bot be found or accessed.
.RE
.sp
.ne 8
.nf
.PP
.B "int MFBClose()"
.fi
.DT
.RS
.I MFBClose
will flush the output buffer and release control of the graphics device
driver.  If the graphics device is a
.I tty,
it is returned to the state that existed prior to the respective
.I MFBOpen
call.
MFBOK is returned if the device was successfully closed;
MFBBADOPT is returned if an error was encountered while attempting
to access or set the device status, and
MFBBADDEV is returned if the output device could not be closed.
.RE
.sp
.ne 12
.nf
.PP
.B "int MFBHalt()"
.fi
.DT
.RS
.I MFBHalt
will flush the output buffer and release control of the graphics device
driver.  If the graphics device is a
.I tty,
it is returned to the state that existed prior to the respective
.I MFBOpen
call.
.I MFBHalt
differs from
.I MFBClose
in that the memory occupied by the respective, current
.I MFB
data structure is not freed.  By calling
.I MFBInitialize,
the graphics device will be reinitialized.  This routine is typically
used by an application program for handling the SIGTSTP signal (the keyboard
stop signal, usually control-Z ).
MFBOK is returned if the device was successfully returned to its initial state;
MFBBADOPT is returned if an error was encountered while attempting
to access or set the device status.
.RE
.sp
.ne 10
.SH SETTING DEVICE PARAMETERS
.PP
Each of the following routines for setting device parameters returns a
diagnostice integer that is defined in the
.I mfb.h
file as follows:
.PP
.ne 10
.nf
.ta \w'XXXXXXXX'u +\w'XXXXXXXXXXXXXXX'u +\w'XXXXXX'u
#define	MFBOK	 1	/* successful return */
#define	MFBBADLST 	-70	/* Illegal line style */
#define	MFBBADFST 	-80	/* Illegal fill style */
#define	MFBBADCST	-90	/* Illegal color style */
#define	MFBBADTM1	-100	/* No destructive text mode */
#define	MFBBADTM2	-110	/* No overstriking text mode */
#define	MFBNOBLNK 	-150	/* No definable blinkers */
#define	MFBTMBLNK 	-160	/* Too many blinkers */
#define	MFBNOMASK	-170	/* No definable read or write mask */
#define	MFBBADALU	-250	/* Cannot set ALU mode */
.fi
.DT
.sp
.ne 11
.nf
.PP
.B "int MFBSetLineStyle(styleId)"
.B "int styleId;"
.fi
.DT
.RS
.I MFBSetLineStyle
sets the current line style to that identified by the integer
.I styleId 
that is greater than or equal to zero and less than the value of
.I maxLineStyles
in the 
.I MFB
data structure.  The value of
.I maxLineStyles
can be obtained from the
.I MFBInfo
routine defined below.  Zero is always the 
.I styleId
for solid lines.  Except for the solid line style,
.I MFB
assumes no default set of lines styles.
MFBOK is returned if the line style was successfully set to that specified by
.I styleId
or if
.I styleId
was already the current line style;
MFBBADLST is returned if
.I styleId
has an illegal value.
.RE
.sp
.ne 11
.nf
.PP
.B "int MFBSetFillPattern(styleId)"
.B "int styleId;"
.fi
.DT
.RS
.I MFBSetFillPattern
sets the current fill pattern to that identified by the integer
.I styleId 
that is greater than or equal to zero and less than the value of
.I maxFillPatterns
in the
.I MFB
data structure.  The value of
.I maxFillPatterns
can be obtained from the
.I MFBInfo
routine defined below.  Solid fill is always defined by
.I styleId
equal to zero.  Other than solid fill,
.I MFB
assumes no default set of fill patterns.
MFBOK is returned if the fill style was successfully set to that specified by
.I styleId
or if
.I styleId
was already the current fill style;
MFBBADFST is returned if
.I styleId
has an illegal value.
.RE
.sp
.ne 11
.nf
.PP
.B "int MFBSetChannelMask(channelMask)"
.B "int channelMask;"
.fi
.DT
.RS
.I MFBSetChannelMask
defines the current write mask to be the value of
.I channelMask.
The channel mask allows specific memory planes to be written and erased
without disturbing other memory planes.  The least significant bit of
.I channelMask
corresponds to the masked value of the first memory plane, etc.
If the corresponding bit is zero, the memory plane is write-protected.
The number of memory planes can be obtained from the
.I MFBInfo
routine defined below.
MFBOK is returned if the write mask was successfully set to
.I channelMask
or if
.I channelMask
was already the current write mask;  MFBNOMASK
is returned if the graphics device does not have a definable write mask.
.RE
.sp
.ne 8
.nf
.PP
.B "int MFBSetReadMask(readmask)"
.B "int readmask;"
.fi
.DT
.RS
.I MFBSetReadMask
defines the current read mask to be the value of
.I readmask.
The read mask allows only specific memory planes to be read.
MFBOK is returned if the read mask was successfully set to
.I readMask
or if
.I readMask
was already the current read mask;  MFBNOMASK
is returned if the graphics device does not have a definable read mask.
.RE
.sp
.ne 10
.nf
.PP
.B "int MFBSetColor(colorId)"
.B "int colorId;"
.fi
.DT
.RS
.I MFBSetColor
sets the current foreground color to that identified by the integer
.I colorId
that is greater than or equal to zero and less than the value of
.I maxColors
in the
.I MFB
data structure.  The value of
.I maxColors
can be obtained from the
.I MFBInfo
routine defined below.  There is no default color map in
.I MFB.
MFBOK is returned if the foreground color
was successfully set to that specified by
.I colorId
or if
.I colorId
was already the current foreground color;
MFBBADCST is returned if
.I colorId
has an illegal value.
.RE
.sp
.ne 14
.nf
.PP
.B "int MFBSetTextMode(destructiveBool)"
.B "Bool destructiveBool;"
.fi
.DT
.RS
.I MFBSetTextMode
defines whether subsequent graphics text will overstrike or replace previous
text.  If
.I destructiveBool
is true, the text mode is set to destructive which means that graphic text
will set the background color of the font grid to the color that is
specified by color style zero depending on the currently defined ALU
operation.  Overstriking mode will only set the pixels of the character
font to the current foreground color.
MFBOK is returned if the graphic text writing mode was successfully
set to that specified by
.I destructiveBool
or if
.I destructiveBool
was already the current graphic text writing mode;  MFBBADTM1
is returned if the graphics device does not have a destructive graphic text
mode, and MFBBADTM2
is returned if the graphics device does not have an overstriking graphic text
mode.
.RE
.sp
.ne 9
.nf
.PP
.B "int MFBSetALUMode(alumode)"
.B "int alumode;"
.fi
.DT
.RS
.I MFBSetALUMode
changes the mode by which the graphics display is changed when an area
of the display is over-written.  The four possible modes
are JAM (replace mode), OR, EOR (exclusive OR), and NOR.
The four valid arguments to
.I MFBSetALUMode
are defined in
.I mfb.h
as follows:
.PP
.ne 9
.nf
.ta \w'XXXXXXXX'u +\w'XXXXXXXXXXXXXXX'u +\w'XXXXXX'u
#define	MFBALUJAM 	0	/* set ALU mode to JAM */
#define	MFBALUOR	1	/* set ALU mode to OR */
#define	MFBALUNOR	2	/* set ALU mode to NOR */
#define	MFBALUEOR	3	/* set ALU mode to EOR */
.fi
.DT
.PP
MFBOK is returned if the ALU mode was successfully set to that specified by
.I alumode
or if
.I alumode
was already the current ALU operation;  MFBBADALU
is returned if the graphics device does not have the ALU mode specified by
.I alumode
or if
.I alumode
is an invalid or illegal argument.
.RE
.sp
.ne 8
.nf
.PP
.B "int MFBSetCursorColor(colorId1, colorId2)"
.B "int colorId1, colorId2;"
.fi
.DT
.RS
.I MFBSetCursorColor
sets the graphics cursor to blink between the two colors identified by
.I colorId1
and
.I colorId2.
The constraints on the values for
.I colorId1
and
.I colorId2
are the same as for
.I MFBSetColor
defined above.  The frequency of the blinking cursor is fixed
and can be changed only by modifying the
.I mfbcap
database file.  MFBOK
is returned if the blinking cursor colors were successfully
set to the specified colors.
.RE
.sp
.ne 10
.nf
.PP
.B "int MFBSetRubberBanding(onFlag, X, Y)"
.B "int X, Y;"
.B "Bool onFlag;"
.fi
.DT
.RS
.I MFBSetRubberBanding
enables/disables rubber banding of the pointing device.
If
.I onFlag
is false, then rubber banding is disabled.
When enabled, the center of rubber banding is at
.I X, Y.
Rubber banding is always disabled immediately
after the pointing device has been used.  MFBOK
is returned if the rubberbanding mode was successfully set to that specified by
.I onFlag;
MFBNORBND is returned if the graphics device does not have
rubberbanding in the pointing device.
.RE
.sp
.ne 13
.nf
.PP
.B "int MFBSetBlinker(colorId, red, green, blue, onFlag)"
.B "int colorId;"
.B "int red, green, blue;"
.B "int onFlag;"
.fi
.DT
.RS
.I MFBSetBlinker
enables the color identified by 
.I colorId
to blink between its currently defined color and the color defined by the
.I red, green, blue
combination.  The values of
.I red, green, 
and
.I blue
are normalized to 1000.  If
.I onFlag
is zero, the blinking is disabled.
The number of colors that may be defined as blinkers at any given time
must be less than the value of
.I maxBlinkers
in the
.I MFB
data structure.  The frequency of the blinking colors is fixed
and can be changed only by modifying the
.I mfbcap
database file.
MFBOK is returned if the color specified by
.I colorId
was successfully set to the desired blinking mode;
MFBNOBLNK is returned if the graphics device does not
have blinking VLT layers, and MFBTMBLNK
is returned if there are already too many active blinking layers.
.RE
.sp
.ne 6
.SH DEFINING DEVICE PARAMETERS
.PP
Each of the following routines for defining device parameters returns a
negative value if any difficulty is encountered.  The possible
returned integers are defined in 
.I mfb.h
as follows:
.PP
.ne 6
.nf
.ta \w'XXXXXXXX'u +\w'XXXXXXXXXXXXXXX'u +\w'XXXXXX'u
#define	MFBOK	 1	/* successful return */
#define	MFBNODFLP 	-120	/* No definable line patterns */
#define	MFBNODFFP 	-130	/* No definable fill patterns */
#define	MFBNODFCO 	-140	/* No definable colors */
.fi
.sp 1
.DT
.ne 10
.nf
.PP
.B "int MFBDefineColor(colorId, red, green, blue)"
.B "int colorId;
.B "int red, green, blue;"
.fi
.DT
.RS
.I MFBDefineColor
redefines the VLT entry for the color identified by
.I colorId
to be the color represented by the
.I red, green, blue
combination where
.I red, green,
and
.I blue
are normalized to 1000.
Once the color corresponding to
.I colorId
is redefined, all geometries that were written onto the
display of a frame buffer with
.I colorId
as the current color will become the new color.
MFBOK is returned if the VLT entry for
.I colorId
was successfully defined;
MFBNODFCO is returned if the graphics device does not have a VLT.
.RE
.sp
.ne 10
.nf
.PP
.B "int MFBDefineFillPattern(styleId, BitArray)"
.B "int styleId;"
.B "int *BitArray;"
.fi
.DT
.RS
.I MFBDefineFillPattern
redefines the fill pattern identified by
.I styleId
and returns with
.I styleId
as the current fill style.
.I BitArray
is a pointer to an array of eight integers whose least significant eight
bits represent individual rows in an eight by eight intensity array.
For example, a fill pattern with an ascending diagonal line may be defined
by the following eight (decimal) integers:
.IP
.ne 6
.nf
.ce
1 2 4 8 16 32 64 128 256

.fi
.DT
.PP
A diagonal-grid fill pattern can be defined with the following integer array:
.IP
.nf
.ce
257 130 68 40 40 68 130 257

.fi
.ne 5
.PP
MFBOK is returned if the new fill style for
.I styleId
was successfully defined;  MFBNODFFP
is returned if the graphics device does not have definable fill patterns.
.DT
.RE
.sp
.ne 9
.nf
.PP
.B "int MFBDefineLineStyle(styleId, Mask)"
.B "int styleId;"
.B "int Mask;"
.fi
.DT
.RS
.I MFBDefineLineStyle
defines the line style identified by
.I styleId
to be the pattern contained in the eight least significant bits of
.I Mask
and returns with
.I styleId
as the current line style.
MFBOK is returned if the line style for
.I styleId
was successfully defined;  MFBNODFFP
is returned if the graphics device does not have definable line patterns.
.RE
.sp
.ne 7
.SH INPUT\/OUTPUT ROUTINES
.nf
.PP
.B "int MFBUpdate()"
.fi
.DT
.RS
.I MFBUpdate
flushes the internal output buffer to the currently defined output device
and will ignore any write error that may occur.  A call to this routine
is ABSOLUTELY necessary to complete any graphics display sequence.
.I MFBUpdate
returns the number of characters sent to the output graphics device or -1 if
a write error occured.
.RE
.sp
.ne 8
.nf
.PP
.B "int MFBPoint(X, Y, key, button)"
.B "int *X, *Y, *button;"
.B "char *key;"
.fi
.DT
.RS
.I MFBPoint
enables the graphics pointing device and then waits for user input.
If a keyboard key is pressed,
.I MFBPoint
returns with
.I key
containing the character that was pressed.
If the pointing device is pressed,
.I MFBPoint
returns with the identified viewport coordinate
.I X, Y,
the contents of 
.I key
equal to zero, and the button mask of the button that was pushed.
The integer array
.I buttonMask
in the 
.I MFB
data structure contains all possible button mask values that can be
returned. 
.I MFBPoint
returns one of the following values defined in the
.I mfb.h
file:
.PP
.ne 6
.nf
.ta \w'XXXXXXXX'u +\w'XXXXXXXXXXXXXXX'u +\w'XXXXXX'u
#define	MFBOK	 1	/* successful return */
#define	MFBPNTERR 	-230	/* Error in pointing device */
#define	MFBNOPTFT 	-240	/* No pointing format */
#define	MFBNOPNT	-260	/* No pointing device */
.fi
.DT
.RE
.sp
.ne 8
.nf
.PP
.B "char *MFBKeyboard(X, Y, background, foreground)"
.B "int background;"
.B "int foreground;"
.B "int X, Y;"
.fi
.DT
.RS
.I MFBKeyboard
enables the graphics keyboard and waits for user input.  A pointer
to a character buffer containing the keyboard input is returned by
.I MFBKeyboard
when the user presses the return or linefeed key.
Backspace is control-H or the delete character, and control-X or control-U
will kill the line.  Pressing the ESCAPE key will cause
.I MFBKeyboard
to return with a null string in the input character buffer.
.PP
All keyboard input is displayed in the viewport with the lower
left corner at the viewport coordinate
.I X, Y
and is constrained to fit on one line.
.I Background
and
.I foreground
are the background and foreground color styles respectively in which the
keyboard input will be displayed.
.RE
.nf
.sp
.ne 5
.PP
.B "void MFBAudio()"
.fi
.DT
.RS
.I MFBAudio
will ring the bell or alarm on the graphics device.  If the graphics device
does not have a bell, then a control-G will be sent to standard output.
.RE
.sp
.ne 12
.nf
.PP
.B "int MFBPutchar(c)"
.B "char c;"
.PP
.B "int MFBPutstr(cp,nchars)"
.B "int nchars;"
.B "char *cp;"
.PP
.B "int MFBGetchar()"
.fi
.RS
These three routines are used internally by
.I MFB
and typically are not used within an application program.
They are comparable to the
.I stdio
routines having similar names.
.I MFBPutchar
places a character
.I c
in the output buffer.
.I MFBPutstr
inserts a string pointed to by
.I cp
containing
.I nchars
characters into the output buffer.  The
.I nchars
argument is necessary to permit embedded null characters in the output
stream.  The characters remain in the output buffer until the next call to
.I MFBUpdate
or until the contents of the output buffer exceed 4096 characters.
.PP
.I MFBGetchar
returns a single character from the graphics input device.  If the graphics
device does not have a keyboard, input is obtained from
the terminal from which the application program was invoked.
If the graphics device is a 
.I tty,
it should be remembered that it is in CBREAK mode.
.RE
.sp 1
.ne 6
.SH TWO DIMENSIONAL GEOMETRY ROUTINES
.PP
All coordinates that are passed to the following geometry routines
are with respect to the display resolution of the graphics device.
.I MFB
assumes that the lower, left corner of the display is the origin with
an absolute coordinate (0, 0).  All coordinate values are positive integers.
.ne 4
.nf
.PP
.B "void MFBMoveTo(X1, Y1)"
.B "int X1, Y1;"
.fi
.DT
.RS
.I MFBMoveTo
sets the current graphics position to
.I X1, Y1.
No line will be drawn from the old graphics position.
.RE
.sp
.ne 5
.nf
.PP
.B "void MFBDrawLineTo(X1, Y1)"
.B "int X1, Y1;"
.fi
.DT
.RS
.I MFBDrawLineTo
draws a line from the current graphics position to
.I X1, Y1
in the current line style and color.
The current graphics position then becomes
.I X1, Y1.
.RE
.sp
.ne 4
.nf
.PP
.B "void MFBLine(X1, Y1, X2, Y2)"
.B "int X1, Y1, X2, Y2;"
.fi
.DT
.RS
.I MFBLine
draws a line in the current line style and color from
.I X1, Y1
to
.I X2, Y2.
.RE
.sp
.ne 4
.nf
.PP
.B "void MFBBox(left, bottom, right, top)"
.B "int left, bottom, right, top;"
.fi
.DT
.RS
.I MFBBox
displays a rectangle in the current fill pattern and color
with diagonal coordinates at
.I left, bottom
and
.I right, top.
.RE
.sp
.ne 12
.nf
.PP
.B "void MFBDrawPath(path)"
.B "MFBPATH *path;"
.fi
.DT
.RS
.I MFBDrawPath
draws a path of vectors in the current line style and color.
.I Path
is a pointer to a data structure defined in the
.I mfb.h
file as follows:
.PP
.nf
.ta \w'XXXXX 'u +\w'XXXXXXXXXXXXXX 'u
struct mfbpath {
	int nvertices;	/* number of (x,y) coordinate pairs */
	int *xy;	/* pointer to array of (x,y) coordinates */
	};

typedef struct mfbpath MFBPATH;
.fi
.DT
.PP
The contents of the coordinate array are organized such that xy[0]
is the x coordinate of the first vertex, xy[1] is the y coordinate of
the first vertex, xy[2] is the x coordinate of the second vertex, etc.
.RE
.sp
.ne 4
.nf
.PP
.B "void MFBFlood()"
.fi
.DT
.RS
.I MFBFlood
erases a frame buffer display to the current color as previously defined by
.I MFBSetColor.
The result would be the same as drawing a solid box over the entire display.
.RE
.sp
.ne 4
.nf
.PP
.B "void MFBPixel(X, Y)"
.B "int X, Y;"
.fi
.DT
.RS
.I MFBPixel
sets the pixel at location
.I X, Y
on the display to the current color as previously defined by
.I MFBSetColor.
.RE
.sp
.ne 7
.nf
.PP
.B "void MFBCircle(X, Y, rad, nsides)"
.B "int X, Y, rad, nsides;"
.fi
.DT
.RS
.I MFBCircle
draws the perimeter of a circle in the current line style and color
with center at
.I X, Y
and with radius
.I rad.
The argument
.I nsides
is the number of line segments with which the circle will be drawn if the
frame buffer does not have a circle primitive.  The default value for
.I nsides
is twenty.
.RE
.sp
.ne 6
.nf
.PP
.B "void MFBFlash(X, Y, rad, nsides)"
.B "int X, Y, rad, nsides;"
.fi
.DT
.RS
.I MFBFlash
draws a round flash with the current fill pattern and color with center at
.I X, Y
and radius
.I rad.
The argument
.I nsides
is the number of line segments with which the flash will be drawn.
The default value for
.I nsides
is twenty.
.RE
.sp
.ne 5
.nf
.PP
.B "void MFBArc(X, Y, rad, angle1, angle2, nsides)"
.B "int X, Y, rad;"
.B "int angle1, angle2, nsides;"
.fi
.DT
.RS
.I MFBArc
draws an arc in the current line style and color with center at
.I X, Y
and with radius
.I rad
beginning at
.I angle1
with respect to the positive y\-axis and ending at
.I angle2
in a counter-clockwise direction.
Both angles are in degrees and are greater than or equal to
zero and less than or equal to 360.  The argument
.I nsides
is the number of line segments with which a 360 degree arc will be drawn.
The default value for
.I nsides
is twenty.
.RE
.sp
.ne 12
.nf
.PP
.B "void MFBPolygon(poly)"
.B "MFBPOLYGON *poly;"
.fi
.DT
.RS
.I MFBPolygon
draws a polygon with the current fill pattern and color.
.I Poly
is a pointer to a data structure defined in the
.I mfb.h
file as follows:
.PP
.nf
.ta \w'XXXXX 'u +\w'XXXXXXXXXXXXXX 'u
struct mfbpolygon {
	int nvertices;	/* number of (x,y) coordinate pairs */
	int *xy;	/* pointer to array of (x,y) coordinates */
	};

typedef struct mfbpolygon MFBPOLYGON;
.fi
.DT
.PP
The contents of the coordinate array are organized such that xy[0]
is the x coordinate of the first vertex, xy[1] is the y coordinate of

lib/news

Type "help" for more information, "quit" to leave.



src/lib/fte/mfb.c
/**********
Copyright 1990 Regents of the University of California.  All rights reserved.
Author: 1988 Jeffrey M. Hsu
**********/

/*
    MFB driver
*/

#include "spice.h"

#ifdef HAS_MFB

/* note:  I would have liked to place this after CPdefs,
    but CPdefs #defines false and true, which are enums in mfb.h. */
#include "mfb.h"

#include "ftedev.h"
#include "cpdefs.h"                     /* for VT_ */
#include "ftegraph.h"           /* for rnd */
#include "suffix.h"

extern char ErrorMessage[];
extern char *MFBHardcopy;	/* output filename for MFB hardcopy */

static int mfbgraphid;
static int mfbscreenflag;
static void mfbsetlinestyles();

static	char old_term[BSIZE_SP] = "";
static	char old_dev[BSIZE_SP] = "";
static  char success = false;

Mfb_Init()
{

    char buf[BSIZE_SP], dev[BSIZE_SP], *term;
    char *devp = dev;
    int err;

    /* Try to identify the terminal type -- check variable,
     * environment, then ask the user.
     */
#ifdef HAS_PCTERM
    term = "pc";
#else
    if (cp_getvar("term", VT_STRING, buf)) {
      term = buf;
    } else {
      term = getenv("TERM");
      if (term == NULL) { 
        /* prompt user w/ input() */
        fprintf(cp_out, "Terminal type = ");
        (void) fflush(cp_out);
        (void) fscanf(cp_in, " %s", buf);
        term = copy(buf);
      }
    }
#endif
    if ((devp = getenv("DEVICE")) == NULL) {
        if (!cp_getvar("device", VT_STRING, dev))
          dev[0] = '\0';
    } else {
	strcpy(dev, devp);
    }
    /* hack for MFB hardcopy */
    if (MFBHardcopy) {
      strcpy(dev, MFBHardcopy);
      if (!strncmp(MFBHardcopy, "/tmp", 4))     /* click to hardcopy */
        mfbscreenflag = 1;
      else
        mfbscreenflag = 0;                      /* hardcopy command */
      MFBHardcopy = (char *) NULL;
    } else {
        mfbscreenflag = 2;                      /* MFB plot */
    }

/* We only re-open if things have changed */
    if( (!success) 
	|| (strcmp(old_term, term) != 0) || (strcmp(old_dev,dev) != 0)) {
    	(void) MFBOpen( term, (dev[0] ? dev : (char *) NULL), &err);
        strcpy( old_term, term); strcpy( old_dev, dev );
        if (err == MFBOK) {
	  success = true;
        } else {
          success = false;
          sprintf(ErrorMessage, "MFB error: %s", MFBError(err));
          externalerror(ErrorMessage);
          return (1);
        }
    }

    dispdev->numlinestyles = MFBInfo(MAXLINESTYLES);
    dispdev->numcolors = MFBInfo(MAXCOLORS);
    dispdev->width = MFBInfo(MAXX) + 1;
    dispdev->height = MFBInfo(MAXY) + 1;

#ifdef DEFCOLORS
    if (dispdev->numcolors > 2 && MFBInfo(DEFCOLORS))
      mfbsetcolors();
#endif
    if (dispdev->numlinestyles > 2 && MFBInfo(DEFLINEPATTERN))
      mfbsetlinestyles();

/*  MFBSetALUMode(MFBALUJAM);   /* graphics overstrikes */
    MFBSetTextMode(1);      /* text overstrikes */

    /* get control back */
    if (MFBHalt() != MFBOK) {
      externalerror("Can't restore control of tty.");
      return(1);
    }

    return (0);

}

Mfb_NewViewport(graph)
GRAPH *graph;
{
    mfbgraphid = graph->graphid;

    if (!MFBInitialize() == MFBOK) {
      externalerror("Can't initialize MFB.");
      return(1);
    }

    if (graph->absolute.width) {
	/* hardcopying from the screen */
	mfbscreenflag = 1;
    }

/* note: for now, do nothing
    later, divide screen into n+1, and redraw existing n plots */

    graph->fontwidth = MFBInfo(FONTWIDTH);
    graph->fontheight = MFBInfo(FONTHEIGHT);

    graph->viewport.height = dispdev->height;
    graph->viewport.width = dispdev->width;

    graph->absolute.xpos = 0;
    graph->absolute.ypos = 0;
    graph->absolute.width = dispdev->width;
    graph->absolute.height = dispdev->height;
    return(0);

}

Mfb_Close()
{

    /* In case of hardcopy command destroy the hardcopy graph
     * and reset currentgraph to graphid 1, if possible
     */
    MFBClose();

    if (!mfbscreenflag) {
	/* DestroyGraph(mfbgraphid); */
	SetGraphContext(1);
    }
    if (mfbscreenflag != 2)
	Mfb_Init();
    mfbscreenflag = 2;

}

Mfb_Clear()
{

    MFBSetColor(0);
    MFBFlood();

}

Mfb_DrawLine(x1, y1, x2, y2)
int x1, y1, x2, y2;
{

    MFBLine(x1, y1, x2, y2);

}

Mfb_Arc(x0, y0, radius, theta1, theta2)
int x0, y0, radius;
double theta1, theta2;
{
	double theta3, theta4;

/* note: figure some better value for nsides as a fn of radius
        and delta theta */
    if( (theta1 != theta2)
		 &&  sin(fabs( theta1 - theta2 )/2.0) < 0.5  ) {
		theta3 = ceil((double) 180.0/M_PI * theta1);
                theta4 = floor((double) 180.0/M_PI * theta2);

        if( theta3 < theta4  && theta3 < 360.0 ) {
    	   MFBArc(x0, y0, radius, (int)theta3, (int)theta4, 50);

	   theta3 *= M_PI/180.0;
	   theta4 *= M_PI/180.0;
	} else {
	   theta3 = theta2;
	   theta4 = theta1;
        }

    	MFBLine((int)(x0+radius*cos(theta1)),(int)(y0+radius*sin(theta1)), 
    	   (int)(x0+radius*cos(theta3)),(int)(y0+radius*sin(theta3)));
    	MFBLine((int)(x0+radius*cos(theta2)),(int)(y0+radius*sin(theta2)), 
    	   (int)(x0+radius*cos(theta4)),(int)(y0+radius*sin(theta4)));
    } else {
    	MFBArc(x0, y0, radius, 
		rnd((double) 180.0/M_PI * theta1),
                rnd((double) 180.0/M_PI * theta2), 50);
    }


}

rrnd( x )
double	x;
{
	if( x - floor( x ) >= 0.5 )
		return( (int)ceil( x ) );
	else
		return( (int)floor( x ) );
}


Mfb_Text(text, x, y)
char *text;
int x, y;
{

    MFBText(text, x, y, 0);

}

Mfb_DefineColor(colorid, red, green, blue)
int colorid;
double red, green, blue;
{

/* note: */
    MFBDefineColor(colorid,
            rnd(red * 1000), rnd(green * 1000), rnd(blue * 1000));

}

Mfb_DefineLinestyle()
{

/* note: */

}

Mfb_SetLinestyle(linestyle)
int linestyle;
{

/* note: do error checking */
    if (MFBSetLineStyle(linestyle) != MFBOK) {
      externalerror("Bad linestyle");
    }

}

Mfb_SetColor(colorid)
int colorid;
{

    if (MFBSetColor(colorid) != MFBOK) {
      externalerror("bad color");
    }

}

Mfb_Update()
{

    MFBUpdate();

}

Mfb_MakeMenu()
{

/* note: */

}

Mfb_MakeDialog()
{

/* note: */

}

/* Set colors properly. Assume that we won't need more than 20 colors. 
 * Also we should try to set the foreground and background colors.
 * These color values are just off of the top of my head...
 * Make the grid color (color 1) blue.
 */

#ifdef DEFCOLORS

static int reds[] = { 999,   0, 999,   0,   0, 999, 999,   0, 999, 500,
              500, 999, 999, 500, 999, 999, 600, 300, 600, 300 } ;
static int grns[] = { 999,   0,   0, 999,   0, 999,   0, 999, 500, 999,
              500, 999, 500, 999, 600, 300, 999, 999, 300, 600 } ;
static int blus[] = { 999,   0,   0,   0, 999,   0, 999, 999, 500, 500,
              999, 500, 999, 999, 300, 600, 300, 600, 999, 999 } ;

static
mfbsetcolors()
{
    int i;
    int red, green, blue;

    MFBDefineColor(1, 0, 0, 999);
    for (i = 2; (i < dispdev->numcolors) && (i < 22); i++) {
        red = reds[i - 2];
        green = grns[i - 2];
        blue = blus[i - 2];
        MFBDefineColor(i, red, green, blue);
    }
    return;
}

#endif

/* Set line styles if necessary. Assume we won't use more than 20 line
 * styles. These line styles also are pretty random. Make sure that
 * line style 0 is solid and line style 1 is dotted.
 */

static int styles[] = { 0333, 0347, 0360, 0071, 0216, 0366, 0155, 0011,
            0123, 0317, 0330, 0145, 0275, 0016, 0315, 0101,
            0222, 0174, 0251, 0376 } ;

static void
mfbsetlinestyles()
{
    int i, j;

    MFBDefineLineStyle(0, 0377);    /* Solid. */
    MFBDefineLineStyle(1, 0125);    /* Dotted. */
    for (i = 2; (i < dispdev->numlinestyles) && (i < 22); i++) {
        j = styles[i - 2];
        MFBDefineLineStyle(i, j);
    }
    return;
}

#endif

